Idempotency is not enough (#2)
exactly-once/Home: content/docs/idempotency-is-not-enough.md

# Introduction
Modern messaging infrastructures offer delivery guarantees that make it non-trivial to build distributed systems. Robust solutions require a good understanding of what can and can't happen in a system and how that affects business level behavior.

This post walks through scenarios that look at the main challenges from the system consistency perspective and sketches possible solutions. 
 
# A system 

We will assume that systems in focus consist of endpoints, each owning a distinct piece of state. Every endpoint processes input messages, modifying its internal state and producing output messages. All endpoints communicate using persistent messaging with at-least-once delivery guarantee. Finally, we will assume that no messages can be lost.

> This covers a pretty wide range of systems. Most notably service-based architectures build on top of modern messaging infrastructure - both on-prem and in the cloud [^1]. 

Reviewer: ```suggestion
This covers a pretty wide range of systems. Most notably service-based architectures build on top of modern messaging infrastructure - both on-premise and in the cloud [^1]. 
```

{{< figure src="/docs/an_endpoint.jpg" title="An endpoint">}}

> With at-least-once delivery, any in-flight message gets delivered possibly multiple times. This is a direct consequence of communication protocols used as any message will be re-deliver until it gets acknowledged by the receiver. Duplicates are created also when messages are produced. A producer can't assume that a message has been delivered and stored by the infrastructure until that gets acknowledged. 

Reviewer: ```suggestion
With at-least-once delivery, any in-flight message gets delivered possibly multiple times. This is a direct consequence of communication protocols used as any message will be re-delivered until it gets acknowledged by the receiver. Duplicates are also created on the producer side as the producer needs to retry sending messages until it gets an acknowledgement from the infrastructure. 
```

Apart from being duplicated, in-flight messages can get re-ordered. There are many reasons for this to happen [^2] one of the most obvious being message re-delivery mechanism. If delivery fails, a message is available for reprocessing only after some back-off period. Any other in-flight message can be processed during that time causing the respective order of those messages to change.

> Duplication and re-ordering happen both at the same which makes any duplication and re-ordering of in-flight messages possible. 

Reviewer: ```suggestion
When combined, duplication and re-ordering can produce, at the receiver side, any sequence of messages. The only guarantee is that the resulting sequence contains at least one copy of each message sent. 
```

TL;DR; Achieving idempotency with business logic code is non-trivial and fragile. And even if done correctly, is not enough to build roboust distributed systems.

## Introduction

> Idempotency can be a useful property when building distributed systems. It's easier said than done however to write idempotent code and keep it so under changing requirements. Finally, generic idempotency is not enough to deal with basic characteristics of communication in distributed systems.

Reviewer: ```suggestion
Idempotency can be a useful property when building distributed systems. It's easier said than done, however, to write idempotent code and keep it so under changing requirements. Finally, generic idempotency is not enough to deal with basic characteristics of communication in distributed systems.
```

## Unboxing idempotency

> There are couple definitions of idempotency but the one we will use here states that: 

Reviewer: ```suggestion
There are a couple of definitions of idempotency but the one we will use here states that: 
```

> An operation `f` is idempotent if `f(f(x)) = f(x)`. 

Put differently, this means that no matter how many times the operation is applied on input `x` the outcome is the same as if the operation was performed only once. 

> In order to get some intuition let's figure out what `f` and `x` stand for in practice. To cover different flavours of practical architectures let's assume that our system consists of number of endpoints. Each endpoint owns distinct piece of state and the only way for the endpoints to communicate is by sending messages. Every endpoint processes input messages one by one modifying it's internal state and producing new messages (1).

Reviewer: ```suggestion
In order to get some intuition let's figure out what `f` and `x` stand for in practice. To cover different flavours of practical architectures let's assume that our system consists of a number of endpoints. Each endpoint owns a distinct piece of state and the only way for the endpoints to communicate is by sending messages. Every endpoint processes input messages one by one, modifying it's internal state and producing new messages (1).
```

*Diagram*: endpoint with state that processes a message multiple times

In this context `f` stands for business logic executed when a message gets processed. `x` on the other hand, represents all resources modified during business logic execution as well as all messages produced. 

## Shoot the target

> It's worth going through a trivial example to see what does idempotency means in a more concrete settings. Let's imagine that our system models a moving target shooting range. It consists of a single `ShootingRange` endpoint which stores shooting target's location `{int: ShootingTarget}` and processes `FireAt : { int: Target }` messages. Whenever `FireAt` message gets processed the endpoint produces either `Hit` or `Missed` event to indicate the result.

Reviewer: ```suggestion
It's worth going through a trivial example to see what does idempotency mean in a more concrete settings. Let's imagine that our system models a moving target shooting range. It consists of a single `ShootingRange` endpoint which stores shooting target's location `{int: ShootingTarget}` and processes `FireAt : { int: Target }` messages. Whenever `FireAt` message gets processed the endpoint produces either `Hit` or `Missed` event to indicate the result.
```

## Target moves (fragile idempotency)

Idempotency is quite fragile and easy to break - even in our trivial system. Let's imagine that we decided to change the target's position with each `FireAt` event. We could start off setting the new position to be the same as `Target` value in the message. Is this still idempotent? Sure. The content of input message is immutable to the new position will always be the same. 

What if we tweaked the logic this way:

``` C#
void Handle(FireAt message)
{
    if(this.ShootingTarget == message.Target) 
    {
        this.ShootingTarget = new Rand().Next(MaxLocation);

        Publish(new Hit());
    }
    else
    {
        Publish(new Missed());
    } 
}
```

Now whenever the target gets hit we choose the new position randomly. Not and overly fancy change, yet enough to break idempotency. The indeterminism introduced by `new Rand().Next(MaxLocation)` makes each re-processing of the same message likely to result in different `ShootingTarget` position.

> What this shows is an instance of a broader scenario. Whenever the logic depends on values outside of our control that can change between exectusion the operation is no longer idempotent. This includes not only random numbers, but things like Guids, local clock values, results of queries to the local file system or remote APIs. If you depend on time or guid identifiers you need to make then deterministics (1). 

Reviewer: ```suggestion
What this shows is an instance of a broader scenario. Whenever the logic depends on values outside of our control that can change between executions, the operation is no longer idempotent. This includes not only random numbers, but things like GUIDs, local clock values, results of queries to the local file system or remote APIs. If you depend on time or GUID identifiers you need to make then deterministic (1). 
```

## Leader board (idempotency but with ids)

Let's extend our system with a second endpoint i.e. `LeaderBoard` that's responsible for storing the number of unique target hits for the player. New endpoint will subscribe and process `Hit` events generated by the `ShootingRange`:

``` C# 
void Handle(Hit message)
{
    this.NumberOfHits++;
}
```

This operation is idempotent but is far from enough in light of possible message duplication. Duplicated `FireAt` messages will result in duplicated `Hit` messages that are indistinguishable from the `LeaderBoard` perspective. We end-up with all operations being idempotent yet with a serious flaw in our system.

> What we need is ability to recognize duplicates at the business level and to be able to that we need some business level identify for each one of them. In other words we need logical message identifiers to be able to cope with duplicates at the logic level. 

Reviewer: ```suggestion
What we need is ability to recognize duplicates at the business level and, to be able to that, we need some business level identity for each one of them. In other words we need logical message identifiers to be able to cope with duplicates at the logic level. 
```

This can be done if we assume that `FireAt` message gets extends with e.g. `AttemptId` property that than gets passed to the `Hit` event. With that in place `LeaderBoard` logic becomes:

``` C#
void Handle(Hit message)
{
    if(this.Hits.Contains(message.AttemptId))
    {
        return;
    }

    this.Hits.Add(message.AttemptId);
    this.NumberOfHits++;
}
```

> What this trivial scenario shows is that generic idempotency is not enough to deal with message duplicates. We need business level identifiers to deal with duplicated messages.

Reviewer: Hmmm what if the business domain does not have a concept of `AttemptId`?
